Breaking: fix config loading (fixes #11510, fixes #11559, fixes #11586)

mysticatea opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[X] Bug fix: fixes #11510, fixes #11559, fixes #11586
[X] Other, please explain: [the refactoring of configurations](https://github.com/eslint/rfcs/blob/42cd563d23eaaad5e9206d8ca7463aeb7dc75a69/designs/2019-eslintrc-improvements/README.md)

Closes #10928 because I realized that this PR does the same thing as https://github.com/eslint/eslint/pull/10928#issuecomment-445324188.

**What changes did you make? (Give an overview)**

This PR fixes three bugs along with the refactoring of configurations.
I'm sorry that this diff is huge.

All new files are in [`lib/cli-engine/`] directory because only `CLIEngine` uses those.

The refactoring does the following items:

- **It simplifies the internal structure of configuration files by an array.**
    - The internal structure is expressed by the classes in [`lib/cli-engine/config-array/`] directory.
    - [`lib/cli-engine/config-array-factory.js`] creates the internal structure from each config file.
        - [This gist](https://gist.github.com/mysticatea/14deeee23dc44b8d7b650d8b9aedc697) is the config array of [our `.eslintrc.js`](https://github.com/eslint/eslint/blob/c021117915d5d23399233f761a237e138f1854af/.eslintrc.js).
    - [`lib/cli-engine/cascading-config-array-factory.js`] handles cascading of config files and CLI options.
- **The config object owns loaded plugins and parsers rather than registers those to somewhere.**
    - [`lib/cli-engine/config-array/config-array.js`] owns loaded plugins and parsers.
    - [`lib/cli-engine/config-array-factory.js`] loads plugins and parsers.
    - [`Linter#getRules()`](https://github.com/eslint/eslint/blob/issue11510/lib/linter.js#L1121) returns the map that contains built-in rules, user-defined rules (by `Linter#defineRule()`), and plugin rules of `ConfigArray` instance the last `Linter#verify()` call used.
    - [`CLIEngine#getRules()`](https://github.com/eslint/eslint/blob/issue11510/lib/cli-engine.js#L594) returns the map that contains built-in rules and plugin rules of `ConfigArray` instances the last `CLIEngine#executeOnFiles()` or `CLIEngine#executeOnText()` call used.
- **It changes the processing order to "config-then-files" from "files-then-config".**
    - [`lib/cli-engine/file-enumerator.js`] finds target files by processing glob patterns. While iterating files, it loads the config file of each directory before iterate files on the directory, so we can use the config files to determine target files. (But, it doesn't use the config files to determine target files yet.)

I describe file mapping of before/after the refactoring. There are details for each new file on the top of the source code.

Old | New
:----|:----
`lib/config-ops.js`<br>(`merge` method) | [`lib/cli-engine/config-array/config-array.js`]
(n/a) | [`lib/cli-engine/config-array/config-dependency.js`]
(n/a) | [`lib/cli-engine/config-array/extracted-config.js`]
`lib/config-ops.js`<br>(`pathMatchesGlobs` method) | [`lib/cli-engine/config-array/override-tester.js`]
`lib/config.js` | [`lib/cli-engine/cascading-config-array-factory.js`]
`lib/config/config-file.js`<br>`lib/util/file-finder.js` | [`lib/cli-engine/config-array-factory.js`]
`lib/util/glob-utils.js`<br>`lib/util/glob.js` | [`lib/cli-engine/file-enumerator.js`]

Therefore, appropriate tests to the moved logics are moved to new files. 

- Tests in `tests/lib/config.js` moved to [tests/lib/cli-engine/cascading-config-array-factory.js#L88](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine/cascading-config-array-factory.js#L88). It's implemented in the `CascadingConfigArrayFactory#getConfigArrayForFile()` method.

- Tests for `applyExtends()`, `load()`, and `resolve()` in `tests/lib/config/config-file.js` moved to [tests/lib/cli-engine/config-array-factory.js#L1216](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine/config-array-factory.js#L1216). It's implemented in the `ConfigArrayFactory` class.

- Tests for `load()` and `loadAll()` in `tests/lib/config/plugins.js` moved to [tests/lib/cli-engine/config-array-factory.js#L2031](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine/config-array-factory.js#L2031). It's implemented in the `ConfigArrayFactory` class.

- Tests for `merge()` in `tests/lib/config/config-ops.js` moved to [tests/lib/cli-engine/config-array/config-array.js#L349](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine/config-array/config-array.js#L349). It's implemented in the `ConfigArray#extractConfig()` method.

- Tests for `pathMatchesGlobs()` in `tests/lib/config/config-ops.js` moved to [tests/lib/cli-engine/config-array/override-tester.js#L101](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine/config-array/override-tester.js#L101). It's implemented in `OverrideTester#test()` method.

- Tests for `listFilesToProcess()` in `tests/lib/util/glob-utils.js` moved to [tests/lib/cli-engine/file-enumerator.js#L171](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine/file-enumerator.js#L171). It's implemented in the `FileEnumerator#iterateFiles()` method.

- Tests for `resolveFileGlobPattern()` in `tests/lib/util/glob-utils.js` moved to [tests/lib/cli-engine.js#L3412](https://github.com/eslint/eslint/blob/issue11510/tests/lib/cli-engine.js#L3412). The new `FileEnumerator` no longer uses this logic, but it's a public API of `CLIEngine`. So we have to keep it.

And I added some `@typedef` comments for VSCode IntelliSense.

**Is there anything you'd like reviewers to focus on?**

- Are there unintentional breaking changes?

I found an additional breaking change (as a bug fix) about plugin loading. Currently, even if a configuration has no `plugins` field, `CLIEngine#executeOnFiles()` and `CLIEngine#executeOnText()` can use plugin rules that the previous calls loaded. After this PR, `CLIEngine#executeOnFiles()` and `CLIEngine#executeOnText()` will report error messages consistently if plugins were not found in configuration.

Related to that, `Linter#getRules()` and `CLIEngine#getRules()` will change:

- `Linter#getRules()` will return the rules, including built-in rules, user-defined rules, and only the plugin rules that the last `Linter#verify()` loaded.
- `CLIEngine#getRules()` will return the rules, including built-in rules and only the plugin rules that the last `CLIEngine#executeOnFiles()` or `CLIEngine#executeOnText()` loaded.

I think that this change is good because formatters can access to unrelated rules (via RFC 10) before this PR and we already have a getter to get information of the last `Linter#verify()` call; `Linter#getSourceCode()`.

[`lib/cli-engine/`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine
[`lib/cli-engine/config-array/`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/config-array
[`lib/cli-engine/config-array/config-array.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/config-array/config-array.js
[`lib/cli-engine/config-array/config-dependency.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/config-array/config-dependency.js
[`lib/cli-engine/config-array/extracted-config.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/config-array/extracted-config.js
[`lib/cli-engine/config-array/override-tester.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/config-array/override-tester.js
[`lib/cli-engine/cascading-config-array-factory.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/cascading-config-array-factory.js
[`lib/cli-engine.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine.js
[`lib/cli-engine/config-array-factory.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/config-array-factory.js
[`lib/cli-engine/file-enumerator.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/file-enumerator.js
[`lib/cli-engine/ignored-paths.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/ignored-paths.js
[`lib/cli-engine/hash.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/hash.js
[`lib/cli-engine/lint-result-cache.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/lint-result-cache.js
[`lib/cli-engine/load-rules.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/load-rules.js
[`lib/cli-engine/module-resolver.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/module-resolver.js
[`lib/cli-engine/naming.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/naming.js
[`lib/cli-engine/xml-escape.js`]: https://github.com/eslint/eslint/tree/issue11510/lib/cli-engine/xml-escape.js

platinumazure commented:
@mysticatea PR #11388 has been merged. It also looks like this needs a rebase/merge. Once that's done, feel free to request my review or ping me and I'll review this. Thanks!

mysticatea commented:
Hi. I updated this PR and I think ready for review!

mysticatea commented:
I updated this PR to resolve conflicts.

mysticatea commented:
I will revert moved files because I have found that Git didn't recognize `cli-engine.js` was moved.

mysticatea commented:
Done.

platinumazure commented on tests/lib/cli-engine/config-array/config-dependency.js:
```diff
@@ -0,0 +1,92 @@
+        });
+    });
+
+    describe("'console.log(...)' should print readable string; not include 'Minimatch' objects", () => {
```
Do we use Minimatch objects in config-dependency?

platinumazure commented on tests/lib/cli-engine/config-array/override-tester.js:
```diff
@@ -0,0 +1,254 @@
+            const tester = OverrideTester.create(files, excludedFiles, basePath);
+
+            assert.strictEqual(
+                JSON.stringify(tester),
```
I think it would be better to test `.toJSON()` and check for presence or absence of properties in the object, but I don't feel strongly.

platinumazure commented on tests/lib/cli-engine/file-enumerator.js:
```diff
@@ -0,0 +1,454 @@
+                });
+            });
+
+            describe("if 'lib/*.js' snf 'test/*.js' were given,", () => {
```
```suggestion
            describe("if 'lib/*.js' and 'test/*.js' were given,", () => {
```

mysticatea commented on lib/cli-engine.js:
```diff
@@ -744,12 +874,16 @@ class CLIEngine {
-        const configHelper = this.config;
-
-        return configHelper.getConfig(filePath);
+    getConfigForFile(filePath = "a.js") {
```
```js
    getConfigForFile(filePath = "a.js") {
        const { configArrayFactory, options } = internalSlotsMap.get(this);
        const absolutePath = path.resolve(options.cwd, filePath);
```

Because:

- that `path.resolve` requires string arguments, 
- there are some tests that call `getConfigForFile()` method with no arguments, 
- and I felt odd to use a directory path to test `overrides` matching.

However, indeed, this will change `overrides` resolving when no arguments. I will make it back.

mysticatea commented on lib/cli-engine/cascading-config-array-factory.js:
```diff
@@ -0,0 +1,393 @@
+ * @typedef {Object} CascadingConfigArrayFactoryOptions
+ * @property {Map<string,Plugin>} [additionalPluginPool] The map for additional plugins.
+ * @property {ConfigData} [baseConfig] The config by `baseConfig` option.
+ * @property {ConfigData} [cliConfig] The config by CLI options. This is prior to regular config files.
```
I wanted to say that CLI options overwrite the setting in config files.

platinumazure replied:
That makes sense, thanks!

mysticatea commented on tests/lib/cli-engine.js:
```diff
@@ -2883,44 +2894,151 @@ describe("CLIEngine", () => {
-    });
 
-    describe("getConfigForFile", () => {
+        describe("config in a config file should prior to shareable configs always; https://github.com/eslint/eslint/issues/11510", () => {
```
Yes, you are right.

mysticatea commented on tests/lib/cli-engine.js:
```diff
@@ -3291,6 +3409,111 @@ describe("CLIEngine", () => {
             });
         });
 
+        describe("Moved from tests/lib/util/glob-utils.js", () => {
```
Actually, this is inside of `describe("resolveFileGlobPatterns", () => {...`

mysticatea commented on tests/lib/cli-engine/cascading-config-array-factory.js:
```diff
@@ -0,0 +1,1224 @@
+             * @returns {void}
+             * @private
+             */
+            function assertConfigsEqual(actual, expected) {
```
Originally, [`assertConfigsEqual` tests only a part of configs](https://github.com/eslint/eslint/blob/a55913d6c6fd1a7c684b8b4d7ab380cf7dc83eb8/tests/lib/config.js#L37). I changed it stricter; the properties in `expected` exist in `actual` with the same values, other properties don't exist in `actual` or exists in `actual` with the default values.

So I should update this JSDoc comment.

mysticatea commented on tests/lib/cli-engine/config-array-factory.js:
```diff
@@ -0,0 +1,2185 @@
+            it("should interpret parser path when present in a JavaScript file", () => {
+                const { ConfigArrayFactory } = defineConfigArrayFactoryWithInmemoryFileSystem({
+                    files: {
+                        "
```
<details><summary>Looks like a bug of GitHub :D</summary>

![](https://user-images.githubusercontent.com/1937871/55996715-2ccdbd80-5cf3-11e9-9d12-da2f0f3f410a.png)

</details>

platinumazure replied:
I see the same thing... I tried to apply it to one of the `describe` calls, probably a few lines above or below.

mysticatea replied:
It's L2031, so this comment slid about 500 lines. I'm not sure why it happened.

mysticatea commented:
Thank you very much for the review!

I updated this PR: [ea25be2...6ea6c1e](https://github.com/eslint/eslint/compare/ea25be2...6ea6c1e?w=1)
I think it addressed those suggestions.

About the Architecture documentation, actually, this PR doesn't change the range of [architecture.md](https://eslint.org/docs/developer-guide/architecture) description. I'm considering to change the basic structure and update `architecture.md` in #11555.

platinumazure commented:
New changes generally look good to me (I do have a few minor nits but I don't have time to write them now- will do later this weekend). I would prefer that more people review this as well.

not-an-aardvark commented on lib/cli-engine/config-array-factory.js:
```diff
@@ -0,0 +1,894 @@
+        let request, id;
+
+        if (isFilePath(nameOrPath)) {
+            request = id = nameOrPath;
```
If I'm understanding correctly, would this allow the `plugins` array to contain a filepath? I think we should disallow this for now until we decide on something like `localPlugins`. (It could cause some ambiguity between slashes in a plugin name and slashes in a rule name.)

mysticatea replied:
Ah, yes, this `if` statement is unnecessary. Actually, the factory rejects file paths before it calls `this.loadPlugin()`.

not-an-aardvark commented on lib/cli-engine/config-array-factory.js:
```diff
@@ -0,0 +1,894 @@
+        try {
+
+            // Resolve the plugin file relative to the project root.
+            const relativeTo = path.join(cwd, ".eslintrc");
```
```suggestion
            const relativeTo = path.join(cwd, "__cwd-placeholder.js__");
```

not-an-aardvark commented on lib/cli-engine/config-array/config-array.js:
```diff
@@ -0,0 +1,445 @@
+        return;
+    }
+
+    for (const key of Object.keys(source)) {
```
Should we check for `__proto__` here? If `source` or `target` has a `__proto__` property, it could cause strange behavior when we set the `__proto__` property of `target`.

not-an-aardvark commented on lib/cli-engine/config-array/config-array.js:
```diff
@@ -0,0 +1,445 @@
+ * @param {Record<string, RuleConf>|undefined} source The source to merge.
+ * @returns {void}
+ */
+function mergeRules(target, source) {
```
Can we rename this to `mergeRuleConfigs`? I was confused at first because I thought `target` and `source` would contain rule objects.

not-an-aardvark commented on lib/cli-engine/config-array/config-array.js:
```diff
@@ -0,0 +1,445 @@
+        delete: { configurable: true, value: void 0 },
+        set: { configurable: true, value: void 0 }
+    });
+}
```
Is the purpose of this just to avoid accidental mutations? I think someone could still mutate a map like this by calling `Map.prototype.set.call(map, foo, bar)`.

mysticatea replied:
> Is the purpose of this just to avoid accidental mutations?

Yes. This `Map` object is private stuff, so this is casual. I wanted to declare that the `Map` object was frozen and avoid copying for performance.

not-an-aardvark commented on lib/cli-engine/config-array/config-array.js:
```diff
@@ -0,0 +1,445 @@
+     * Check if this config has `root` flag.
+     * @type {boolean}
+     */
+    get root() {
```
Would it be possible to change these to methods instead of getters? It seems like using getters could cause confusing performance behavior if someone expects them to just be normal properties.

not-an-aardvark commented on lib/cli-engine/config-array/override-tester.js:
```diff
@@ -0,0 +1,180 @@
+            return b;
+        }
+
+        return new OverrideTester(a.patterns.concat(b.patterns), a.basePath);
```
Is using `a.basePath` correct here? It seems like this will cause `b.patterns` to be resolved against the wrong path.

mysticatea replied:
Actually, the `basePath` property of `a` and `b` are the same value always. I will add `assert`.

not-an-aardvark commented on lib/config/config-initializer.js:
```diff
@@ -573,7 +553,16 @@ function promptUser() {
+            } else if (config.extends) {
+                config.extends = [config.extends, earlyAnswers.styleguide];
+            } else {
+                config.extends = earlyAnswers.styleguide;
```
Maybe it would be better to use an array here so that it's more clear how the user can extend additional configs if needed.

```suggestion
                config.extends = [earlyAnswers.styleguide];
```

not-an-aardvark commented on lib/config/config-rule.js:
```diff
@@ -299,8 +296,8 @@ function generateConfigsFromSchema(schema) {
 function createCoreRuleConfigs() {
-    return Object.keys(builtInRules).reduce((accumulator, id) => {
-        const rule = rules.get(id);
+    return Array.from(builtInRules.keys()).reduce((accumulator, id) => {
```
Could this be:

```js
return Array.from(builtInRules.entries()).reduce((accumulator, [id, rule]) => {
    // ...
});
```

Then it wouldn't be necessary to define `rule` on the following line.

not-an-aardvark replied:
Actually, I realized that this could just be 

```js
return Array.from(builtInRules).reduce((accumulator, [id, rule]) => {
    // ...
});
```

But it's not a blocker.

not-an-aardvark commented on lib/config/config-validator.js:
```diff
@@ -141,20 +144,26 @@ function validateRuleOptions(rule, ruleId, options, source = null) {
+function validateEnvironment(
+    environment,
+    source,
+    getAdditionalEnv = Function.prototype
```
Can we use something like

```js
const noop = Function.prototype;

function validateEnvironment(
    environment,
    source,
    getAdditionalEnv = noop
) {}
```

instead? I think the usage of `Function.prototype` directly as a noop function might be confusing without further explanation.

not-an-aardvark commented on lib/config/config-validator.js:
```diff
@@ -164,17 +173,23 @@ function validateEnvironment(environment, envContext, source = null) {
+function validateRules(
+    rulesConfig,
+    source,
+    getAdditionalRule = Function.prototype
```
For clarity, can we use a variable called `noop` here rather than `Function.prototype` directly?

not-an-aardvark commented on lib/linter.js:
```diff
@@ -343,24 +351,23 @@ function normalizeVerifyOptions(providedOptions) {
-
-    const mergedParserOptions = ConfigOps.merge(parserOptionsFromEnv, providedOptions || {});
-
+        .reduce((parserOptions, env) => lodash.merge(parserOptions, lodash.cloneDeep(env.parserOptions)), {});
```
Is it necessary to clone `env.parserOptions` here? I thought `lodash.merge` doesn't mutate the second argument.

mysticatea replied:
I wanted to prevent `lodash.merge` modifies `env.parserOptions.nestedProperty`. But, indeed this `lodash.cloneDeep` was unnecesary.

mysticatea commented on lib/cli-engine.js:
```diff
@@ -154,105 +193,62 @@ function calculateStatsPerRun(results) {
-    });
-    const stats = calculateStatsPerFile(fixedResult.messages);
+    const basename = path.basename(filePath, path.extname(filePath));
+    const resultFilePath = basename.startsWith("<") && basename.endsWith(">")
```
Yes.

`ConfigArray#extractConfig(filePath)` method requires an absolute path, but `Linter` object doesn't know CWD. So CLIEngine passes an absolute path always. If a user didn't specify the file path on `CLIEngine`'s methods, `filePath` is an absolute path like `/path/of/cwd/<text>`. On the other hand, the lint result uses only `<text>` part.

not-an-aardvark replied:
Could we explicitly check for `<text>` instead (e.g. when converting to an absolute path)? It seems like this could be a confusing edge case if someone has files that start with `<` and end with `>` for some reason.

mysticatea commented on lib/cli-engine.js:
```diff
@@ -406,6 +436,72 @@ function getCacheFile(cacheFile, cwd) {
+            return map;
+        }, {});
+    }
+    return void 0;
```
This returns undefined for the optimization in the below `createConfigDataFromOptions` function.

mysticatea commented on lib/cli-engine.js:
```diff
@@ -406,6 +436,72 @@ function getCacheFile(cacheFile, cwd) {
+        plugins === void 0 &&
+        rules === void 0
+    ) {
+        return null;
```
CLI options become a `ConfigArray` element. This is a small optimization to not make the `ConfigArray` element if there is no option.

not-an-aardvark replied:
It seems like no `ConfigData` will be created if a user's `.eslintrc.js` file just contains `module.exports = {};`. Will that cause a "No ESLint Configuration" error in that case?

mysticatea replied:
No. This handles only CLI options. Config files make one or more `ConfigArray` element always.

mysticatea commented on lib/cli-engine.js:
```diff
@@ -587,92 +664,139 @@ class CLIEngine {
+        const extensions = options.extensions.map(ext => ext.replace(/^\./u, ""));
+        const dirSuffix = extensions.length === 1
+            ? `/**/*.${extensions[0]}`
+            : `/**/*.{${extensions.join(",")}}`;
```
Actually, this is copied from [lib/util/glob-utils.js#L57-L63](https://github.com/eslint/eslint/blob/bc0819c94aad14f7fad3cbc2338ea15658b0f272/lib/util/glob-utils.js#L57-L63). I'm OK for the simplification.

mysticatea commented on lib/cli-engine.js:
```diff
@@ -744,12 +874,17 @@ class CLIEngine {
-        return configHelper.getConfig(filePath);
+        const { configArrayFactory, options } = internalSlotsMap.get(this);
+        const absolutePath =
+            filePath ? path.resolve(options.cwd, filePath) : options.cwd;
```
This is for backward compatibility. Several tests have called `getConfigForFile` method without arguments...

not-an-aardvark replied:
I think it would be reasonable to require the argument since this is a breaking change anyway and calling it with no arguments is undocumented. But I don't feel strongly about it.

mysticatea replied:
This was my memory's confusion; tests without arguments were for old `configHelper.getConfig(filePath)`. I have removed this check.

mysticatea commented on lib/cli-engine.js:
```diff
@@ -782,7 +916,8 @@ class CLIEngine {
 
-            const cwd = this.options ? this.options.cwd : process.cwd();
+            const slots = internalSlotsMap.get(this);
+            const cwd = slots ? slots.options.cwd : process.cwd();
```
Yes. When `CLIEngine.getFormatter()` static method was called, it's undefined.

mysticatea commented on lib/cli-engine/cascading-config-array-factory.js:
```diff
@@ -0,0 +1,391 @@
+            name: "--rulesdir",
+            filePath: "",
+            plugins: {
+                "": new ConfigDependency({
```
Indeed it's hacky, but [lib/cli-engine/config-array/config-array.js#L267](https://github.com/eslint/eslint/blob/45f724ebc8067e8b0650fe5d4dc43653b3d0c9c7/lib/cli-engine/config-array/config-array.js#L267) handles empty string as special a bit; it doesn't add `/`.

mysticatea commented on lib/cli-engine/cascading-config-array-factory.js:
```diff
@@ -0,0 +1,391 @@
+        if (configArray.length > 0) {
+            configArray.unshift(...parentConfigArray);
+        } else {
+            configArray = parentConfigArray;
```
Indeed, because `ConfigArray` inherits `Array`, it's mutable. But `ConfigArray#extractConfig()` method makes cache, so mutation after `ConfigArray#extractConfig()` was called causes bugs. Currently, `ConfigArray` is private and mutation happens only in initialization.

Probably we should freeze those for our maintenance.

mysticatea commented:
Thank you very much for making time to review this large PR!

I think that I fixed all the suggested changes.

not-an-aardvark commented:
This looks good to me after https://github.com/eslint/eslint/pull/11546#discussion_r282655081 is addressed.

Also, I think this PR should remove [these lines](https://github.com/eslint/eslint/blame/bc0819c94aad14f7fad3cbc2338ea15658b0f272/docs/user-guide/migrating-to-6.0.0.md#L183) from the migration guide.

mysticatea commented:
Thank you!
I updated this PR for those suggestions.

mysticatea commented:
Thank you very much for the review of this huge PR!
